### DISCUSSION TITLE
Questions about development

### DISCUSSION THREAD

**Discussion by twistios** [fast4x/RiMusic#2589] [General]
I am opening this discussion for questions regarding code contributions, pull requests and so on.

**Comment by twistios** [comment]
In `extensions/nextvisualizer/painters/fft` Android Studio complains there that `org.apache.commons` is unresolved.
It seems that this can be fixed by including `implementation 'org.apache.commons:commons-math3:3.6.1'` (or something similar).
It looks to me that in this project it is solved differently. Also everything does compile, but still it is annoying that it shows this error.
How can I solve this problem?

**Comment by fast4x** [comment]
go to gradle/libs.....toml and check if math3 is present. After sync gradle library

**Reply by twistios** [reply]
It was present in the `libs.versions.toml`-file and the problem was fixed by doing a gradle sync. But I think that I already had done multiple gradle syncs before.

**Comment by fast4x** [comment]
update your repo ..

**Comment by twistios** [comment]
How can one translate the "encrypted" source file names to regular file names? @fast4x

For example the following stack trace:
````
XXXX-XX-XXTXX:XX:XX.XXX: java.io.IOException: Too many open files
sun.nio.ch.IOUtil.makePipe(Native Method)
sun.nio.ch.PollSelectorImpl.(PollSelectorImpl.java:57)
sun.nio.ch.PollSelectorProvider.openSelector(PollSelectorProvider.java:36)
y5.a.v(SourceFile:38)
d7.a.i(SourceFile:9)
v7.J.run(SourceFile:95)
A7.i.run(SourceFile:239)
C7.k.run(SourceFile:3)
C7.a.run(SourceFile:96)
````

**Comment by twistios** [comment]
How long is `currentVisualizerKey` stored (or actually the value that is stored)? Across versions? I am very unexperience with Android-Kotlin.

**Reply by fast4x** [reply]
Is an Int

**Reply by twistios** [reply]
I was able to answer my own question I guess. The information is stored across different runs of the app so I guess it is stored permanently and also across versions of the app. Ok, so this means I am going to open a pull request about this.

**Reply by twistios** [reply]
https://github.com/fast4x/RiMusic/pull/2726

**Comment by twistios** [comment]
Is it expected to have `ArrayIndexOutOfBoundsException` from 
````
//PlayerModern
val prevPrevMediaItem = try {
            binder.player.getMediaItemAt(prevMediaItemIndex - 1)
        } catch (e: Exception) {
            MediaItem.EMPTY
        }
````
multiple times (per second) in a "normal scenario"? The Index is `-1` as `prevMediaItemIndex` is `0`. The current song was song number 1 (counting from 0) in the queue.

@fast4x you did not have to close this discussion. Or should I open a new one for every question?

**Reply by fast4x** [reply]
I don't understand what's your purpose

**Reply by twistios** [reply]
For me it looks like there is no problem (saying that I found none until yesterday) when instead of `prevMediaItemIndex - 1` it is only `prevMediaItemIndex` AND there is no exception regularly happening.

**Comment by twistios** [comment]
Do you also have the problem, that in `compose/routing` none of the files can finish analysing in AndroidStudio? Not even the 15 line `OnGlobalRoute` can finish after 5 minutes (it does not finish).

I guess this is also part of the reason why there are errors when importing from `compose/routing` in `HomeScreen.kt`

[Screencast from 2024-07-24 18-10-20.webm](https://github.com/user-attachments/assets/1d6f8c9b-99af-420f-8dd3-4f235e3ecf8c)

**Reply by twistios** [reply]
I installed AndroidStudio 2024.1.2 Beta 1 (because of another problem) and now indexing and imports work fine again.

**Comment by twistios** [comment]
I have found some bug in the app that I want to fix but because of a bug ([youtrack](https://youtrack.jetbrains.com/issue/IDEA-326539/Debugger-stuck-on-Processing-classes-for-emulated-method-breakpoints)) in AndroidStudio (which in reality is NOT resolved and [Google](https://issuetracker.google.com/issues/331834527#comment30) and Jetbrains (see youtrack-link) blame each other for a part of it), I can not easily debug this issue in the app.

**Reply by twistios** [reply]
I guess I need to open another issue then on youtrack for the part where it seems to be an issue with IntelliJ IDEA.
